fix(a11y): main menu a11y updates

bbsmooth opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes locally on my machine.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #45106

<!-- Feel free to add any additional description of changes below this line -->

This is a big one. There are several a11y issues with the current main menu that I have attempted to fix. 

- The language picker is basically inaccessible for keyboard users. In order to keep it in the main menu I had to create a custom submenu. This is by far the majority of new code/changes in this PR.
- The keyboard focus indicator is not consistent, and sometimes not even visible across browsers. I really feel that FCC needs to settle on a universal focus indicator color. For the time being I have chosen the `--blue-mid` color from the predefined color pallet for the menu because it has accessible contrast with both dark and light themes and I just think it looks nice. You'll also notice that you will only see the focus outline if you are using the keyboard to navigate the menu. The ability to do this easily with CSS has been added rather recently (and has a safe fallback for those browsers who don't recognize it).
- Converted the menu to a list as that is the current best practice and it allows screen reader users to know how many items are in the menu.
- Moved the avatar/sign-in button out of the menu. These are not visually located in the menu and thus should be outside of it. Also, the avatar appears to go to the same place as the Profile menu option, but that's for another discussion.
- Some other little niceties for keyboard users such as auto-closing the menu when it loses focus and allowing the Escape key to close the menu.

I was not able to figure out how to get the various languages to serve properly, so I was not able to do an official test on the language picker to make sure it takes you to the language you pick. I would appreciate it if someone would test this for me. Also, I don't have access to Safari, so it would be great if someone who does would test this, especially the language picker, to make sure it behaves properly.

This PR is not quite ready to go as-is but it is definitely good enough for preliminary testing. There are some translations that will need to be set up. And I'm not a React expert (yet), so if I have done some things that go against best practices I'm more than happy to hear about it. Same with TypeScript. And I'm sure not every change I made will be perfect.

bbsmooth committed:
fix: main menu a11y updates

nhcarrigan commented:
> I was not able to figure out how to get the various languages to serve properly, so I was not able to do an official test on the language picker to make sure it takes you to the language you pick

For what it's worth, the different languages in production are entirely different instances - there's not really a way to test switching them locally without rebuilding the entire client.

ahmadabdolsaheb commented:
Thanks for the hard work here. Here are a few nitpicks.

- Everything in the right side of the navigation bar should be in Lato. (Except the sign in cta button)
- The sign in button in navigation is not centered. (For signed out users)
- Keep the navigation section dividers to group items by functionality.

In order to be more conventional with the lang selector, I suggest having a span with the word “Language:”  followed by “English” and aria-label of “Change language" or “Select Language”.
Here is an example in the [footer](https://www.khanacademy.org).

I completely agree on implementing a consistent focus state. And —blue mid would be a great candidate for the reason you mentioned above.

After completing the ui library, a color system audit is due where we evaluate our accent colors and their usage.

For the time being, let us have the focus state represented as an `outline: 3px solid var(--blue-mid);`
For the submenu, we could go anywhere between gray-85 to gray-75 for the background, a similar hover and focus state as the main menu.

For navigating between clients, if the new language selector links to a similar url as before, it should be good to go.

Other than that the navigation is coming tougher neatly here.

bbsmooth committed:
update font menu

bbsmooth commented:
The font in the menu is currently set to `Roboto-Mono, sans-serif` but since `Roboto-Mono` isn't an actual font name (it should be `Roboto Mono` without the dash) then the font is falling back to `sans-serif`. I have changed my PR to use `sans-serif` but if this is not the actual intended font then let me know what it should be.

bbsmooth committed:
fix: sign in button text alignment

bbsmooth committed:
fix: disabled button font

bbsmooth committed:
fix: language menu font

bbsmooth committed:
remove sign-in from main menu

bbsmooth committed:
fix: add section dividers to menu

bbsmooth committed:
fix: height on language selector menu

bbsmooth committed:
fix: alignment of Menu button text

bbsmooth commented:
> In order to be more conventional with the lang selector, I suggest having a span with the word “Language:” followed by “English” and aria-label of “Change language" or “Select Language”.

This is actually a violation of [WCAG 2.5.3 Label in Name](https://www.w3.org/WAI/WCAG21/Understanding/label-in-name.html). (Yes, Khan Academy is in violation of this guideline.) This guideline says that the accessible name (which would be the text in `aria-label` in this case) must include the text in the visible label for the control. So we can't have the visible text for the language picker be something like "Language: English" and then set the accessible name (using `aria-label`) to something like "Change language". 

The way I have it implemented now, there is no need for `aria-label` because the accessible name is just the text in the button ('Change Language'). Personally, I didn't think it was necessary to show the current language since the language would be shown in that language's font. So for example, if you can't read Chinese, and the page is currently in Chinese, then I'm not sure what good it would do to have the language picker actually show the language name since only people who can read Chinese would understand it, and they probably don't need to be told what language they are currently reading.

I think what would be more useful would be to find an icon that represents a language picker (I'm not sure what that would be at the moment) and then if somebody accidentally switches to a language they can't read at least they would be able to find the language picker easily by recognizing the icon.

bbsmooth commented:
> * Everything in the right side of the navigation bar should be in Lato. (Except the sign in cta button)
> 
> * The sign in button in navigation is not centered. (For signed out users)
> 
> * Keep the navigation section dividers to group items by functionality.

I have fixed these issues and also a few minor touch-ups of issues I overlooked initially. Eagerly awaiting more feedback. If everyone feels that the language picker needs to show the current language then I will work on making that as accessible as possible.

ahmadabdolsaheb commented:
I agree with your reasoning, let's keep it as "change language".

I think the globe icon is a good candidate. We might consider replacing it with the drop down icon. 
As of the styles, let's do Latto for everything in the dropdown menu and the language menu.

On a second thought, let's change the background of the language menu to a lighter gray shade( between gray-85 to gray-75), and keep the hover state on nav menu and language menu identical to what it was previously.
We are rethinking our buttons' states, and will roll the changes across platforms once we finish the v1 of our component library. 

This seems to be coming together nicely. Thanks a lot for your efforts.

bbsmooth committed:
fix: language globe icon

bbsmooth committed:
refactor: remove dup selector

bbsmooth commented:
I searched for "language icon" and found a possible candidate which I have added to the main menu. Let me know if you like it or if you want me to keep trying. 

As for the colors on the language menu, FCC is currently using a `<select>` for this menu and I don't believe you can style it much with CSS, so the colors you are seeing are determined by the OS/browser. I tried to mimic the colors used on Windows/Linux the best I could. I think the blue hover is pretty close. I could make the background a slightly darker shade of gray but I wouldn't go too much darker or we'll have to turn the text white. Personally, I think the contrast of the light background compared with the dark background of the main menu makes it stand out better, but I'm not a designer so take that with a grain of salt. Let me know if you still want any changes to it.

ahmadabdolsaheb commented:
Everything looks great. 

Here are the final touches:
- font of the toggle menu and the language menu should be Lato.
- background color of the language menu should be var(--gray-80); 
- color of the language menu should be same as the toggle menu.
- checkmark color should be the same as the font color
- The globe should replace the dropdown icon.
- hover color/background of the ui elements should stay as the following for consistency:
```css
color: var(--theme-color);
background-color: var(--gray-10);
```
(We will incorporate that blue background further in the component library.)

Here are a few quick mocks. 

<img width="298" alt="Screen Shot 2022-04-01 at 2 55 58 PM" src="https://user-images.githubusercontent.com/4591597/161268741-1d02e0f8-dc4d-4edd-9a51-e14bff2ade25.png">


<img width="301" alt="Screen Shot 2022-04-01 at 2 58 11 PM" src="https://user-images.githubusercontent.com/4591597/161268657-6961dcb6-4423-459e-aa0a-3b0d090f060b.png">
(the checkmark is missing)

bbsmooth committed:
update language menu colors

bbsmooth commented:
@ahmadabdolsaheb Thank you for the mock-ups, this made it much clearer as to what you want. I believe I got it all but let me know if I missed something. If this looks good to go I believe we just need to add a translation for "Change Language".

ahmadabdolsaheb commented:
In addition to adding the ` menuButtonRef: {} as Ref<HTMLButtonElement>, hideMenu: jest.fn() ` to the `LandingPageProps` in the `header.test.tsx` needs to be restructured.

Since the tests are tightly coupled with the previous structure of the navigation components, I wonder if they should be replaced with cypress tests instead.

ahmadabdolsaheb commented on client/src/components/Header/index.tsx:
```diff
@@ -71,9 +88,13 @@ export class Header extends React.Component<
           <UniversalNav
             displayMenu={displayMenu}
             fetchState={fetchState}
+            displayLanguageMenu={displayLanguageMenu}
```
nitpick: we might need to change `displayLanguageMenu` to `isLanguageMenuDisplayed` or something so it is not confused with `showLanguageMenu`.

bbsmooth committed:
refactor: clearer name for language menu display state

bbsmooth committed:
fix: don't close onBlur if Menu button is clicked

bbsmooth commented:
@ahmadabdolsaheb Good call on changing the name of `displayLanguageMenu`. I also fixed a small minor annoyance concerning the onBlur event closing the menu too early if the focus was on one of the last two items in the main menu. Only keyboard users who first put the focus on one of those items and then used their mouse to close the menu by clicking the main Menu button would have noticed it, but I wouldn't be able to sleep at night if I left it in there.

ojeytonwilliams commented on client/src/components/Header/components/nav-links.tsx:
```diff
@@ -16,33 +16,43 @@ import {
 
 const locales = availableLangs.client;
 
+const languageGlobeIconStyle = {
+  height: '1rem',
+  width: '1rem'
+};
```
We use LanguageGlobe once, right?  If so, we can just fold this into the component.

bbsmooth replied:
> If I tab outside the menu it closes. Is this necessary? It seems like that would be an issue if a user accidentally presses tab too often.

Ya, this is one of those "gray" areas when it comes to accessibility. The nav menu is technically a "disclosure" widget. The W3C ARIA Authoring Practices actually has an example of a [disclosure pattern for navigation menus](https://w3c.github.io/aria-practices/examples/disclosure/disclosure-navigation.html). If you look at note 3 under "Accessibility Features" it says:

_If a dropdown is open and focus is inside the navigation region, pressing Esc will close the dropdown. **Moving focus out of the navigation region also closes an open dropdown.**_

That's the pattern I was following here but I can completely understand reasons for not doing this. I think one of the main reasons I lean toward autoclose is because with narrower view ports (narrow enough that it shows the tab interface) the dropdown menu covers the tabs completely, and at some point the menu covers the entire top half of the page (or more, depending on the view port height). So that could potentially lead to users losing sight of the focus indicator because it is hidden underneath the menu, which I think an argument could be made that it would then violate [WCAG SC 2.4.7](https://www.w3.org/WAI/WCAG21/quickref/?currentsidebar=%23col_customize&levels=aaa#focus-visible). 

A possible compromise? The next tab stop after the last menu item is the profile icon. We could keep the menu expanded if the user tabs to that icon, which would allow them one "free" tab, and then close the menu if they tab off of the icon into the content.

ojeytonwilliams replied:
That's fair enough.  It's a little annoying to have to shift-tab back and try to navigate again if you accidentally go too far, but it's not *that* bad.  I think it's fine to err on the side of following the w3c examples.

We could go with the compromise, but the code is already a little on the complicated side. I'd rather keep it simple for now.  If we get feedback from users that this is frustrating, we can revisit this, but it seems good for now.

bbsmooth replied:
Leaving as is for now.

ojeytonwilliams commented on client/src/components/Header/components/nav-links.tsx:
```diff
@@ -64,23 +86,198 @@ export class NavLinks extends Component<NavLinksProps, {}> {
+    if (!currentListItem) {
+      return menuButtonRef.current;
+    }
+    try {
+      const previousSibling =
+        currentListItem.previousSibling as HTMLElement | null;
+      const previousListItem = previousSibling.querySelector('a, button');
+      return previousListItem;
+    } catch (e) {
+      return menuButtonRef.current;
+    }
```
Why is this a try catch block?  It seems like this is using error handling to handle normal behaviour.

ojeytonwilliams reviewed (dismissed):
Hey @bbsmooth thanks for all your hard work on this, it is streets ahead of our old menu.  I did notice a couple of minor things. 

If I tab outside the menu it closes. Is this necessary?  It seems like that would be an issue if a user accidentally presses tab too often.

The search bar is inaccessible on mobile, since clicking on it closes the menu.

Also, the Change Language text needs to be translatable.

Finally, on a general note, there are few slightly iffy patterns dotted through this PR.  Switch statements are hard to read and error prone.  [here](https://www.30secondsofcode.org/articles/s/javascript-switch-object) is a good post about how and why to avoid them.  Tangentially related, mutating state is something avoid if possible (i.e. setting and checking `keyFlag` isn't necessary.  Instead you can have the key handlers return the boolean.

bbsmooth committed:
refactor: move globe icon styling to CSS

bbsmooth commented:
> The search bar is inaccessible on mobile, since clicking on it closes the menu.

Good catch, I missed this. After looking into this more, I think the search bar has an issue with [WCAG SC 3.2.3](https://www.w3.org/WAI/WCAG21/Understanding/consistent-navigation.html). The search feature comes before the drop down menu in the DOM and is always visible in wide view. But in narrow view it is hidden unless the menu is expanded, so the user first has to expand the menu and then go backwards in the DOM to access the search feature. In one instance the search is completely independent of the menu and in the other it is dependent on the menu even though its position in the DOM is the same.

I'd like to do some thinking on this and see if I can come up with a solution that provides consistency and gives you the desired UI you currently have in place. If anyone has any suggestions I'm all ears.

bbsmooth committed:
refactor: get rid of switch statements

bbsmooth commented:
Removed the switches. To be honest, the only time I ever use them is with keyboard handlers, they just seem like a perfect fit, and most of the keyboard handlers I've seen over the years have used them. But I'm a man, but I can change, if I have to, I guess. (Bonus props if anyone gets that reference).

bbsmooth committed:
refactor: remove try catch block

bbsmooth committed:
fix: translate Change language button

ojeytonwilliams commented:
Thanks @bbsmooth it's all looking cleaner now. And shorter, too, which is always nice!

Regarding the search menu, I hadn't considered the possibility of keyboard navigation on small screens. Now that I've tried it I see that it's very counter-intuitive to have to tab forwards to the menu, open it and then shift-tab backwards to the search bar.

What do you think to simply putting the search component in the menu on smaller screens and leaving it where it is for larger ones?

bbsmooth commented:
@ojeytonwilliams suggested:
> putting the search component in the menu on smaller screens

So there is actually a WCAG Success Criterion ([3.2.3](https://www.w3.org/WAI/WCAG21/Understanding/consistent-navigation.html)) that addresses this very issue and it basically says that your site search navigation needs to stay in the same place in the DOM. Man, I am a killjoy, aren't I? Good thing I have kittens or you all would probably hate me by now.

But I do have an idea. If we are willing to keep a small search icon at the top and to the left of the FCC logo then we can use that to show the search bar on demand. Behind the scenes, this icon would really be the search input field and setting focus on it would cause the complete search bar to display below the top nav bar (just like it currently does). This would decouple the search bar from the menu completely.

nhcarrigan commented:
> Man, I am a killjoy, aren't I?

Someone's got to keep all my wild ideas in check 🤣

bbsmooth commented:
@ojeytonwilliams You are correct, it is concerned about the consistency between pages. The "out" here is the last few words of the criterion: "unless a change is initiated by the user". I actually thought about this a lot and it basically came down to a particular example. What if a blind user is using a mobile device and at certain times they have it oriented so that it is in wide view and at other times they have it oriented so that it is in narrow view? The blind user is not going to see the change visually or hear that a change occurred but yet the search feature is going to move on them. Initially I was thinking that this would be a good enough example to show that the intent of SC 3.2.2 was being violated. But now that I think about it more, I think this would technically be considered a user initiated change and thus you can forget what I said earlier and we could move the search into the menu on narrow view if we wanted to.

I still think I like the idea of having a consistent icon on the nav bar though, but instead of having it to the left of the FCC logo we could move it next to the Menu button. If we do move it into the menu then we are going to have to make some fixes (which we should probably do anyway). For example, right now it is possible to Tab out of the suggestion list and leave the list showing, which would then cover the other menu items. (This actually already happens if you are in a very narrow view.) Also, we would need a method to close the list quickly (such as the Escape key). I'm not finding an easy way to do this right now but perhaps I just don't know enough about how to use it? I also don't like the idea of having two tab stops for one menu item (the search input field and the search button). But If this is the way everyone wants to go I'm sure we can make it work.

Would you be opposed to me implementing my persistent icon idea real quick, just to see? Or do you think moving it to the menu is probably inevitable?

ojeytonwilliams commented:
> Would you be opposed to me implementing my persistent icon idea real quick, just to see? Or do you think moving it to the menu is probably inevitable?

Nope, this sounds like a good idea, so if you'd like to give it a go, please do!

Nothing is set in stone - I just suggested moving the search widget into the menu as a (relatively) straightforward fix.  It's just one option we can explore.

Sembauke commented:
Hey @bbsmooth

How is this going, do you need any help?

ahmadabdolsaheb commented:
I am leaning towards the single menu button for simplicity; However, As Oliver mentioned, nothing is set in stones. We can evaluate the icon option if a single button does not work.

I noticed that some websites that include the search bar in their dropdown menus deactivate the auto search functionality for mobile, so that is another option.

bbsmooth commented:
@ahmadabdolsaheb:
> I am leaning towards the single menu button for simplicity;

I'm not sure I completely understand what this solution would entail. Could you point me to an example or give a more detailed explanation?

ahmadabdolsaheb commented:
The github dropdown navigation is a good example.  If you are using the browser, refresh the page after toggling the device toolbar so the mobile version shows up.

bbsmooth committed:
fix: move search into nav menu for mobile layout

bbsmooth committed:
fix: forgot a merge

bbsmooth commented:
I've moved the search feature so it comes directly after the menu toggle button when viewing with the mobile layout. I could use a bit of help though. After I rebased to fix some merge conflicts the dev server builds just fine but I'm immediately getting an unhandled runtime error:

can't access property "english", _config_i18n_all_langs__WEBPACK_IMPORTED_MODULE_8__.langCodes is undefined

The console is showing that this is triggered on line 431 of nav-links.tsx. I'm not sure I'm smart enough to figure this one out on my own. Any help would be appreciated.

bbsmooth committed:
refactor: updates for changes in i18n/all-langs

bbsmooth committed:
fix: prevent menu from collapsing when focus is on change language button and user clicks into search

bbsmooth committed:
fix: translate cancel change option in language picker

bbsmooth commented:
> How is this going, do you need any help?

Hey @Sembauke Thanks for the offer and checking up on this. I think I've got this ready to go now. Please test it out and let me know what you think.

ahmaxed committed:
feat: add cypress tests

ahmaxed committed:
feat: display the complete language list

ahmadabdolsaheb commented on client/src/components/Header/components/nav-links.tsx:
```diff
@@ -16,33 +16,39 @@ import {
   toggleNightMode: (x: any) => any;
   user?: Record<string, unknown>;
   navigate?: (location: string) => void;
+  showLanguageMenu?: (elementToFocus: HTMLAnchorElement) => void;
```
I think the argument for many of the functions that operate in the nav-links should be a 'HTMLButtonElement'. 

That being said, there are a lot of TS rules that are being ignored in the navigation related components. I don't mind handling in a whole another pr.

bbsmooth replied:
You are correct. Not sure why I used `HTMLAnchorElement` when all of the items in the language menu are buttons? I guess I was just thinking of the main nav menu in which most of the items are links.

bbsmooth committed:
fix: fix TS typing

bbsmooth committed:
fix: force scrollbar on lang menu

bbsmooth commented:
@ahmadabdolsaheb I think we need the height set to 100% on the lang menu. If you are not logged in then the main menu is shorter than the lang menu and if the height is not set then no vertical scrollbar appears and so it is not obvious that there are more language choices in the menu. With height set to 100% then the vertical scrollbar will appear.

However, when you are signed in then the main menu is longer than language menu but the language menu will extend passed its last item and cover the entire main menu. Was this your intent?

Sembauke commented:
Hey Bruce,

I am not an expert on accessibility.

There seems to be some issue with trying to focus on the elements, this is what I tried:

1. Open the menu
2. Press tab multiple times until you are focused on "change language"
3. Hit Enter
4. Press Tab again

It loses focus when not logged in, it focuses on the login button when you are. 

Here is an example:

https://user-images.githubusercontent.com/46919888/170677478-e989c743-0ed4-49b7-8736-5d1bad4bb0fa.mp4

ahmadabdolsaheb commented:
> However, when you are signed in then the main menu is longer than language menu but the language menu will extend past its last item and cover the entire main menu. Was this your intent?

Depending on the user's OS and browser, the  scrollbar might not be displayed. Here is the current state of the pr on a macOS and chrome:
<img width="657" alt="Screen Shot 2022-05-27 at 4 29 21 PM" src="https://user-images.githubusercontent.com/4591597/170708961-96a5904b-fba7-456c-af7a-c829b15225c5.png">

I see your point about the logged out state. My intent was expanding the `nav-lang-menu` so all of the languages visible without a need of a scrollbar.

bbsmooth commented:
@Sembauke The language menu is designed to work as a "real menu" in which the arrow keys are used to navigate through the language options, not the Tab key. When the language menu is open, the Tab key still works like it does in the main menu in that it takes you to the next item in the main menu (or the Sign In button if you aren't logged in). Most "real menus" work this way. The pattern is that you navigate into/out of the menu with the Tab key but you navigate within the menu with the arrow keys. But I can definitely understand the confusion here. First of all, we are combining a "navigation menu" (which doesn't use the arrow keys) with an actual menu that does use the arrow keys. Second, I originally styled the language menu to look like the current one, which is using a `<select>`, with the thought that people are used to using the arrow keys with select lists and so they would also use them with our language menu. But the language menu is now styled so similar to the main menu that I can see how someone might assume the Tab key would navigate the language menu as well.

Did you eventually figure out that the arrow keys navigate the language menu? I'm working on the assumption that people who regularly use a keyboard to navigate would know to try the arrow keys, but I'm more than willing to admit this is a bad assumption. I'm not opposed to allowing the Tab key to navigate the language menu as well, but there are some subtle issues that would need to be addressed if we did this. I'm wondering if we can't just go with what we have now and then perhaps encourage keyboard users to test the menu and get their feedback?

bbsmooth committed:
fix: remove scroll bar from lang menu

bbsmooth commented:
@ahmadabdolsaheb I think I have a solution for displaying all of the language options without needing a scroll bar, even when the user is not signed in. It does require specifying the height of the language menu in the CSS, so as other languages are added that height will need to be adjusted. I'm guessing there are better options for doing this that would not require changing the CSS but perhaps we could go with this for the time being and work on an alternative in another PR?

bbsmooth committed:
fix: close menu when user tabs away from last menu item

EmmaDawsonDev reviewed (commented):
I think you've made some great improvements and are well on the way to making this much more accessible. I've been testing this out with a screen reader and keyboard and here's a few things that I noticed that need addressing:

- When the menu opens the focus should move to the first item in the list automatically, then when it closes it should move back to the button.
- the menu button should have aria-haspopup="true" as an attribute
- when using arrow keys to navigate up/down through the menu the screen reader announces the next link but the visible focus does not follow (the focus stays on the last tabbed key)
- the new blue focus outline only works on the search bar and menu button, it doesn't seem to apply to the profile button or any other buttons on the page. It would be good if the focus outline were consistent across all elements of all pages.

__
[![sema-logo](https://app.semasoftware.com/img/sema-tray-logo.gif)](https://semasoftware.com/gh) &nbsp;**Summary:** :hammer_and_wrench: This code needs a fix

bbsmooth commented:
Hi @emmalearnscode, thanks for the feedback. The menu is based on a disclosure widget pattern:

bbsmooth commented:
Hi @emmalearnscode Thank you for the feedback. The main menu is based on the disclosure widget pattern, which is pretty standard for site navigation controls. Adding `aria-haspopup` would require that the list have `role="menu"` which I think is discouraged here. Although I have seen some menus automatically put the focus on the first item when opened (microsoft.com) I don't think it is a requirement and I have seen far more leave the focus on the toggle button, so I am tempted to leave it as is. The main menu does not currently support arrow key navigation. The APG for disclosure menus indicates that arrow key support is optional. It might be advantageous to add it but I don't think it should hold up this PR (as it's been long enough in the making).

[APG Example Disclosure Navigation Menu](https://www.w3.org/WAI/ARIA/apg/example-index/disclosure/disclosure-navigation.html)
[Don’t Use ARIA Menu Roles for Site Nav](https://adrianroselli.com/2017/10/dont-use-aria-menu-roles-for-site-nav.html)

As for the focus indicators, I'm not 100% sure it has been settled that the blue I am using is the official color. It is definitely my recommendation and so I used it to demonstrate how well it works. But I think an issue probably needs to be opened on this so FCC can get this settled and all the focus indicators across the site can be updated. Regardless, this PR was just for the menu and so I don't think a global issue like this should hold up this PR.

emmalearnscode commented:
Yes, of course all these things are down to choice. I was looking at this widget instead: https://www.w3.org/WAI/ARIA/apg/example-index/menu-button/menu-button-links.html but they are very similar. As with our other discussion, it's probably best to get real screen reader user feedback in order to determine their preference about the change in focus.

I know the arrow up/down is optional but in this case there is actually a bug (I was testing on chrome, let me know if you can't replicate and I will try to take a video) If you tab to the first item 'Donate' in the list so that it has focus it will be read out by the screen reader as expected. If you then press on the arrow down key the next item 'Curriculum' in the menu will be read out by the screen reader but the focus will stay on 'Donate'. Pressing enter will activate 'Curriculum' so the arrow up/down is working but the focus indicator is not following so either the focus needs to follow the arrow keys too or the arrow keys should not trigger the cycling through the menu items.

bbsmooth commented:
Regarding the APG navigation menu button example, there is a [github issue](https://github.com/w3c/aria-practices/issues/353) about removing it and the navigation menubar example, or at least putting a big warning about using them for anything other than what would be considered a true application menu, that goes back for years. I think the consensus is that aria menu roles should not be used for site navigation. But it just occurred to me, we can't use this pattern even if we wanted to since we are incorporating the search feature into the menu for narrow view ports and there is no aria menu role that we could apply for the search in the menu. So I don't think this even involves testing. We are stuck using the disclosure pattern to accommodate the search.

Regarding the arrow keys, the main nav menu does not support them at all right now. I think when you are using the arrow keys with your screen reader you are in browse mode and thus the arrow keys allow you to jump around the content, but that doesn't force focus to change. In other words, you are using the the functionality provided by the screen reader, not the menu. If you Tab through the menu then you should see the focus indicator follow as usual because Tabbing is supported in the menu.

emmalearnscode commented:
> Regarding the APG navigation menu button example, there is a [github issue](https://github.com/w3c/aria-practices/issues/353) about removing it and the navigation menubar example, or at least putting a big warning about using them for anything other than what would be considered a true application menu, that goes back for years. I think the consensus is that aria menu roles should not be used for site navigation. But it just occurred to me, we can't use this pattern even if we wanted to since we are incorporating the search feature into the menu for narrow view ports and there is no aria menu role that we could apply for the search in the menu. So I don't think this even involves testing. We are stuck using the disclosure pattern to accommodate the search.
> 
> Regarding the arrow keys, the main nav menu does not support them at all right now. I think when you are using the arrow keys with your screen reader you are in browse mode and thus the arrow keys allow you to jump around the content, but that doesn't force the focus outline to follow. In other words, you are using the the functionality provided by the screen reader, not the menu. If you Tab through the menu then you should see the focus indicator follow as usual because Tabbing is supported in the menu.

Thanks for the info about the APG examples. I wish they'd make this information more available for everyone and not expect everyone to go hunting through open github issues for the latest best practices. It's hard enough to learn about accessibility as it is.

I also realised my mistake with the arrow keys when I woke up this morning - lesson learnt -> don't do reviews late at night and mix up keyboard testing and screen reader testing.

ahmadabdolsaheb commented:
@emmalearnscode, thanks for taking a look. Yes we will gradually roll out the focus styles to all of our components in this repository.

moT01 reviewed (commented):
Well, I looked over this pretty hard. There's a lot of conversation here - I tried to read, or at least skim over, all of it - I apologize if I missed some context or explanation - here's everything I found...

I was a little confused with using the tab key then switching to the arrows for the menu, too. I picked up on it quick.

The language links all look to be working.

On Safari 13.1.2 - pretty old version - there's quite a few issues with it. Tab order seems off - I hit enter on the menu and then tab takes me to 'night mode', and 'change language' after that. All the other options get skipped. It also skips over the avatar/profile icon. Well, none of it is even reachable on production with that version of Safari - so it’s an improvement - since that’s the case, I wouldn’t let that hold back this PR. I tried looking into it a little, didn’t see what the cause was. Did anyone else (@ahmadabdolsaheb) get a chance to test this on a newer Safari? I do think this might be something to try and look into with another PR - The nav is heavily used so it might be worth making it more compatible than other things like a lone curriculum test.

There did seem to be a minor style issue with the search on smaller windows...
Production - search lines up with whatever button is above it:
<img width="232" alt="Screen Shot 2022-05-31 at 2 42 46 PM" src="https://user-images.githubusercontent.com/20648924/171289288-a43cc3c7-1cb4-41d3-83df-ebf4210a7b96.png">
This PR - search goes behind the scroll bar:
![Screen Shot 2022-05-31 at 2 43 04 PM](https://user-images.githubusercontent.com/20648924/171289357-90c89fb1-ec69-4107-9a55-bcb45b4a6d96.png)

I left a few minor suggestions below.

That's really all I found? I must be losing it - or this just looks really good. It looks real good @bbsmooth 🎉 Nice job 👍 Here's some quotes from you: "I'm not a React expert" ... "Same with TypeScript." Whatever you say. I would be good with merging it like this if we want.

Edit: This has been around a while and had a lot of work put in. All the things I found are quite minor - if you're ready to be done with it, @bbsmooth - let me know and we can merge it.

bbsmooth committed:
add list role to navigation list to appease Safari

bbsmooth committed:
chore: capitalize Change Language/remove CSS comment

bbsmooth commented on client/src/components/Header/components/universal-nav.css:
```diff
@@ -104,28 +105,36 @@
   min-height: var(--header-height);
   width: 100%;
   border: none;
+  /*line-height: 2.1;*/
```
Whoops. Fixed

bbsmooth commented on client/i18n/locales/english/translations.json:
```diff
@@ -70,7 +70,9 @@
     "click-start-course": "Start the course",
-    "click-start-project": "Start the project"
+    "click-start-project": "Start the project",
+    "change-language": "Change language",
```
Fixed.

bbsmooth commented:
@moT01 
> On Safari 13.1.2 - pretty old version - there's quite a few issues with it. Tab order seems off - I hit enter on the menu and then tab takes me to 'night mode', and 'change language' after that. All the other options get skipped.

Wow, I was not expecting that. I know that there can be issues with the VoiceOver screen reader and Safari if the list has had the default bullet styling removed, but I don't think I've heard of it being a problem when not using VO. The solution for the VO issue is to add `role='list'` to the list, which apparently reminds VO that it is a list even though it doesn't have bullets. On a long shot, I went ahead and added the list role. Can you try it again and see if it made any difference? If not, then I'll have to do some research. But ya, since it doesn't work at all on production at least it's a slight improvement. But I hope this wouldn't hold up this PR, it's been long enough in the making.

bbsmooth commented on client/src/components/Header/components/universal-nav.css:
```diff
@@ -142,33 +151,148 @@
+
+.nav-lang-menu-option[aria-current='true'] {
+  /* check mark for current language */
+  background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' width='32' height='32' preserveAspectRatio='xMidYMid meet' viewBox='0 0 16 16'%3E%3Cg fill='white'%3E%3Cpath d='M12.736 3.97a.733.733 0 0 1 1.047 0c.286.289.29.756.01 1.05L7.88 12.01a.733.733 0 0 1-1.065.02L3.217 8.384a.757.757 0 0 1 0-1.06a.733.733 0 0 1 1.047 0l3.052  ...
```
Ya, perhaps not my finest hour. I think I went with the CSS route because it was the least amount of work. No problem refactoring this but if we could do it in a separate PR that would be great.

bbsmooth committed:
fix: right side search box alignment in narrow view

bbsmooth commented:
@moT01 
> There did seem to be a minor style issue with the search on smaller windows...

You'd think by now all browsers would handle CSS the same but of course this is an instance where it looked good in FF but not the others. Someday I'll learn that you've got to test everything in every browser.

I think I've got this looking good in all of them now.

moT01 commented:
> The solution for the VO issue is to add `role='list'`...

No, that didn't change anything @bbsmooth.

I found [the issue](https://stackoverflow.com/questions/1848390/safari-ignoring-tabindex) though, it was a safari setting.

bbsmooth commented:
> I found [the issue](https://stackoverflow.com/questions/1848390/safari-ignoring-tabindex) though, it was a safari setting.

Oh ya, I completely forgot about that issue. I had heard of it before, but since I don't have a mac it just wasn't one that popped into my head. I'll remove the extraneous list role and I think we are ready to to.

bbsmooth committed:
remove extraneous list role

moT01 committed:
fix: cypress

moT01 reviewed (approved):
LGTM 🎉 

Want to take one last look and get this in when you get up @ahmadabdolsaheb
